source/code/plugins/in_npmd.rb

```diff
@@ -0,0 +1,676 @@
+                    next if _line.nil?
+                    _json = check_and_get_json(_line.chomp)
+                    unless !_json.nil?
+                        Logger::logWarn "Sent string to plugin is not a json string", Logger::loop
+                    else
+                        unless _json.key?("DataItems") and !_json["DataItems"].nil? and _json["DataItems"] != ""
+                            Logger::logWarn "No valid data items found in sent json #{_json}", Logger::loop
+                        else
+                            _uploadData = _json["DataItems"].reject {|x| x["SubType"] == "ErrorLog"}
+                            _errorLogs  = _json["DataItems"].select {|x| x["SubType"] == "ErrorLog"}
+                            _uploadData.each do |item|
+                                if item.key?("SubType")
+                                    # Append FQDN to path data
+                                    if !@fqdn.nil? and item["SubType"] == "NetworkPath"
+                                        item["Computer"] = @fqdn
+                                    # Append agent Guid to agent data
+                                    elsif !@agentId.nil? and item["SubType"] == "NetworkAgent"
+                                        item["AgentId"] = @agentId
+                                    end
+                                end
+                            end
+                            emit_upload_data_dataitems(_uploadData) if !_uploadData.nil? and !_uploadData.empty?
+                            emit_error_log_dataitems(_errorLogs) if !_errorLogs.nil? and !_errorLogs.empty?
+                        end
+                    end
+                rescue StandardError => e
+                    log_error "Got error while reading data from NPMD: #{e}", Logger::loop + Logger::resc
+                end
+            end while !@npmdClientSock.nil?
+        end
+
+        def package_and_send_error_log(msg)
+            _dataItems = Array.new
+            _h = Hash.new
+            _h["Message"] = msg
+            _dataItems << _h
+            emit_error_log_dataitems(_dataItems)
+        end
+
+        def emit_error_log_dataitems(dataitems)
+            _record = Hash.new
+            _record["DataType"] = "HEALTH_ASSESSMENT_BLOB"
+            _record["IPName"]   = "LogManagement"
+            _record["DataItems"] = dataitems
+            router.emit(Tag, Engine.now, _record)
+        end
+
+        def emit_upload_data_dataitems(dataitems)
+            _record = Hash.new
+            _record["DataType"] = "NETWORK_MONITORING_BLOB"
+            _record["IPName"]   = "NetworkMonitoring"
+            _record["DataItems"] = dataitems
+            router.emit(Tag, Engine.now, _record)
+        end
+
+        def wait_for_agent_id
+            if @agentId.nil?
+                begin
+                    @agentId = OMS::Configuration.agent_id()
+                    sleep(0.1)
+                end while @agentId.nil?
```

Comment:
Firstly, the agent id is not necessary in the data uploaded from the agent. It could be enriched in the ODS after authentication.
Secondly, if you really want to insert agent id into the data, you need to initialize the OMS::Configuration in this plugin?